apps/astarte_appengine_api/lib/astarte_appengine_api/rooms/room.ex

```diff
@@ -293,6 +296,47 @@ defmodule Astarte.AppEngine.API.Rooms.Room do
     end
   end
 
+  defp verify_device_exists(realm_name, device_id) do
+    case device_exists?(realm_name, device_id) do
+      true ->
+        :ok
+
+      false ->
+        _ =
+          Logger.warn(
+            "Device #{device_id} in realm #{realm_name} does not exist.",
+            tag: "device_does_not_exist"
+          )
+
+        {:error, :device_does_not_exist}
+    end
+  end
+
+  defp device_exists?(realm_name, encoded_device_id) do
+    with {:ok, decoded_device_id} <- Device.decode_device_id(encoded_device_id),
+         {:ok, client} <- connect_to_db_with_realm(realm_name),
+         {:ok, result} <- check_device_exists(client, decoded_device_id) do
+      result
+    end
+  end
+
+  defp connect_to_db_with_realm(realm_name) do
+    with {:error, :database_connection_error} <- Database.connect(realm: realm_name) do
+      {:error, :database_connection_error}
+    end
+  end
+
+  defp check_device_exists(client, device_id) do
+    with {:error, :database_error} <- Queries.check_device_exists(client, device_id) do
+      Logger.warn(
+        "Database error when checking existence of device #{inspect(device_id)}",
+        tag: "check_device_exists_db_error"
+      )
+
+      {:error, :database_error}
+    end
+  end
```

Comment:
Imho this whole chunk is a little bit confusing, you have `verify_device_exists`, `device_exists?` and `check_device_exists` and it's not clear what's the difference between them without reading the implementation. You should just keep `verify_device_exists` because:
- `check_device_exists` is just catching the error to log, but you're actually already logging inside the function, so you could directly call `Queries.check_device_exists` where you're calling the local `check_device_exists`
- `device_exists?` has a `?` suffix but it doesn't actually return a boolean because it could return `{:error, reason}` or a boolean, which is a strange pattern (usually if you return `{:error, reason}` you return `:ok` or `{:ok, result}`). You should move `device_exists?` body inside `verify_device_exists` and then check result with an `if` (it's a boolean so no need for a `case`).